hw/ip/otbn/doc/dv_plan/index.md

```diff
@@ -24,83 +16,124 @@ applicable. Once done, remove this comment before making a PR. -->
+- A [TileLink interface]({{< relref "hw/dv/sv/tl_agent/README.md" >}}).
+  OTBN is a TL device, which expects to communicate with a TL host.
+  In the OpenTitan SoC, this will be the Ibex core.
+- An idle signal, `idle_o`
+- Two interrupts
+- An [alert interface]({{< relref "hw/dv/sv/alert_esc_agent/README" >}})
+
+The idle and interrupt signals are modelled with the basic
+[`pins_if`]({{< relref "hw/dv/sv/common_ifs#pins_if" >}}) interface.
+
+As well as instantiating OTBN, the testbench also instantiates an `otbn_core_model`.
+This module wraps an ISS (instruction set simulator) subprocess and performs checks to make sure that OTBN behaves the same as the ISS.
+It has a single output signal that goes high when there's a mismatch with the model.
+The testbench wraps this signal with a [`pins_if`]({{< relref "hw/dv/sv/common_ifs#pins_if" >}}) interface.
+
+### otbn_agent
+
+There is a single agent, called `otbn_agent`.
+This is in charge of driving `tl_if`, the TileLink interface to start and stop OTBN.
+It also monitors `alert_if`, `int_if` and `idle_if` for changes in state and `err_if` for any errors reported by the model.
 
 ### Reference models
-[Describe reference models in use if applicable, example: SHA256/HMAC]
+
+The main reference model for OTBN is the instruction set simulator (ISS), which is run as a subprocess by DPI code inside `otbn_core_model`.
+This Python-based simulator can be found at `hw/ip/otbn/dv/otbnsim`.
 
 ### Stimulus strategy
+
+When testing OTBN, we are careful to distinguish between
+
+- behaviour that can be triggered by particular instruction streams
+- behaviour that is triggered by particular external stimuli (register writes; surprise resets etc.)
+
+Testing lots of different instruction streams doesn't really use the UVM machinery, so we have a "pre-DV" phase of testing that generates constrained-random instruction streams (as ELF binaries) and runs a simple block-level simulation on each to check that the RTL matches the model.
+The idea is that this is much quicker for designers to use to sanity-test proposed changes, and can be run with Verilator, so it doesn't require an EDA tool licence.
+This pre-DV phase cannot drive sign-off, but it does use much of the same tooling.
+
+Once we are running full DV tests, we re-use this work, by using the same collection of randomised instruction streams and randomly picking from them for most of the sequences.
+The ELF binaries are all placed in a single directory and named sequentially (`0.elf`, `1.elf`, ..., `1234.elf` or similar).
+This directory and the maximum index are passed to the simulation with plusargs.
+
+<div class="bd-callout bd-callout-warning">
+
+**TODO:** There's a bit of infrastructure work to do here to teach `dvsim.py` to run the instruction generator. Also, maybe it would be better to pass a list of ELF files (which would let us filter the initial generation for things that hit coverage points in pre-DV). This design might change slightly.
```

Comment:
Could this be done entirely in Systemverilog (pass the elf file as a plusarg and some SV logic will invoke readelf (or an external script) to extract stuff. I need to do something like this for the chip level as well. It will be best to sync up and gather requirements.

Reply:
Reading the ELF files isn't too bad. The problem is that we have to generate them at some point. I'm just not sure how this all gets plumbed in so that you can "just run dvsim.py" to trigger everything. I feel like we're probably going to want some configuration file (in one of the existing hjson files, maybe) that says "generate N randomised ELF files (starting at seed S) in directory D" and then we need to make sure that this happens before the first simulation (since we might have `# sims > N` and don't want to duplicate work).

But anyway, I thought I'd just leave a "TBD" note here and fix it up once we've figured out what works best!